modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -270,6 +263,61 @@ private <T> CompletableFuture<T> enlistInTx(
         return postEnlist(fut, implicit, tx0);
     }
 
+    /**
+     * Retrieves a batch rows from replication storage.
+     *
+     * @param tx Internal transaction.
+     * @param partId Partition number.
+     * @param scanId Scan id.
+     * @param batchSize Size of batch.
+     * @return Batch of retrieved rows.
+     */
+    private CompletableFuture<Collection<BinaryRow>> enlistCursorInTx(
+            InternalTransaction tx,
+            int partId,
+            long scanId,
+            int batchSize
+    ) {
+        final boolean implicit = tx == null;
+
+        final InternalTransaction tx0 = implicit ? txManager.begin() : tx;
+
+        String partGroupId = partitionMap.get(partId).groupId();
+
+        ClusterNode clusterNode = tx0.enlistedNode(partGroupId);
+
+        CompletableFuture<Collection<BinaryRow>> fut;
+
+        if (clusterNode != null) {
+            try {
+                fut = replicaSvc.invoke(clusterNode, tableMessagesFactory.readWriteScanRetrieveBatchReplicaRequest()
+                        .groupId(partGroupId)
+                        .transactionId(tx.id())
+                        .scanId(scanId)
+                        .batchCounter(batchSize)
+                        .build());
+            } catch (Throwable e) {
+                throw new TransactionException(format("Failed to enlist rows into a transaction"));
```

Comment:
Please add specific message, now it's about rows instead of scans.

Reply:
Done.